ksqldb-engine/src/main/java/io/confluent/ksql/physical/scalablepush/ScalablePushRegistry.java

```diff
@@ -0,0 +1,188 @@
+import io.confluent.ksql.physical.scalablepush.locator.PushLocator;
+import io.confluent.ksql.schema.ksql.LogicalSchema;
+import io.confluent.ksql.util.PersistentQueryMetadata;
+import io.vertx.core.impl.ConcurrentHashSet;
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.function.Supplier;
+import org.apache.kafka.streams.StreamsConfig;
+import org.apache.kafka.streams.kstream.KStream;
+import org.apache.kafka.streams.kstream.Windowed;
+import org.apache.kafka.streams.processor.Processor;
+import org.apache.kafka.streams.processor.ProcessorContext;
+import org.apache.kafka.streams.processor.ProcessorSupplier;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * This registry is kept with every persistent query, peeking at the stream which is the output
+ * of the topology. These rows are then fed to any registered ProcessingQueues where they are
+ * eventually passed on to scalable push queries.
+ */
+public class ScalablePushRegistry {
+
+  private static final Logger LOG = LoggerFactory.getLogger(ScalablePushRegistry.class);
+
+  private final KStream<?, GenericRow> stream;
+  private final LogicalSchema logicalSchema;
+  private final PushLocator pushLocator;
+  private final Set<ProcessingQueue> processingQueues = new ConcurrentHashSet<>();
+
+  public ScalablePushRegistry(
+      final KStream<?, GenericRow> stream,
+      final LogicalSchema logicalSchema,
+      final PushLocator pushLocator,
+      final boolean windowed
+  ) {
+    this.stream = stream;
+    this.logicalSchema = logicalSchema;
+    this.pushLocator = pushLocator;
+    registerPeek(windowed);
+  }
+
+  public void close() {
+    for (ProcessingQueue queue : processingQueues) {
+      queue.close();
+    }
+  }
+
+  public void register(final ProcessingQueue processingQueue) {
+    processingQueues.add(processingQueue);
+  }
+
+  public void unregister(final ProcessingQueue processingQueue) {
+    processingQueues.remove(processingQueue);
+  }
+
+  public PushLocator getLocator() {
+    return pushLocator;
+  }
+
+  @VisibleForTesting
+  int numRegistered() {
+    return processingQueues.size();
+  }
+
+  // Registers a peek processor which makes a copy of every row and passes it on to any registered
+  // queues.
+  @SuppressWarnings("unchecked")
+  private void registerPeek(final boolean windowed) {
+    final ProcessorSupplier<Object, GenericRow> peek = new Peek<>((key, value, timestamp) -> {
+      for (ProcessingQueue queue : processingQueues) {
+        try {
+          final TableRow row;
+          if (!windowed) {
+            final GenericKey keyCopy = GenericKey.fromList(((GenericKey) key).values());
+            final GenericRow valueCopy = GenericRow.fromList(value.values());
+            row = Row.of(logicalSchema, keyCopy, valueCopy, timestamp);
```

Comment:
can we do this outside of the for loop or am I missing something?

Reply:
thanks for the detailed explanation - totally agree we shouldn't change the original code, it might have some real performance implications elsewhere.